File: 0011_container_with_most_water/container_water.cpp
#include <iostream>
#include <climits>
#include <vector>

using namespace std;

/**
 * Time: O(n)
 * Space: O(1)
 */
int max_area(vector<int> &height)
{
    int l = 0;
    int r = height.size() - 1;
    int maximum = INT_MIN;
    while (l < r)
    {
        int minimum = min(height[l], height[r]);
        maximum = max(maximum, minimum * (r - l));
        height[r] > height[l] ? l++ : r--;
    }
    return maximum;
}

File: 0011_container_with_most_water/container_water.py
from typing import List


def max_area(height: List[int]) -> int:
    """
    Time: O(n)
    Space: O(1)
    """
    maximum = float("-inf")
    l = 0
    r = len(height) - 1
    while l < r:
        minimum = min(height[l], height[r])
        maximum = max(maximum, (r - l) * minimum)
        if height[r] > height[l]:
            l += 1
        else:
            r -= 1
    return maximum
